### ISSUE TITLE
[RFC] Allow using rocksdb as local cache storage

### ISSUE THREAD

**Issue by flaneur2020** [slatedb/slatedb#177]
> it even contains a built-in FIFO strategy which is useful on cache eviction

> Everyone I've spoken to that's doing object store has suggested FIFO RocksDB as well.

_Originally posted by @criccomini in https://github.com/slatedb/slatedb/issues/137#issuecomment-2329536746_

this issue is expected to have some discussion before actual work on this.

rocksdb's BlobDB has a nice FIFO strategy which nicely fit the idea from the paper [FIFO Queues are All You Need for Cache Eviction]( https://dl.acm.org/doi/pdf/10.1145/3600006.3613147). all you need is just store the cached blob into it and without worrying about eviction or any metadata bookkeepings, as BlobDB already handled these for you.

the current design in #139 expects the cached part files not to be very small (like 4kb), and if the part size is small, you'll get massive small files in the fs (if your disk is 1tb, you'll got 0.2 billion 4kb small files (i suspect the modern fs like XFS can handle 1 billion small files properly but needs some tests to verify it)).

rocksdb may help us merge these small files automatically.

to adapt rocksdb as the local cache storage, all what we need is implement this trait in https://github.com/slatedb/slatedb/pull/139/files#diff-c9320be53c7d123e524c26deaf3a89596fa301964fc635815908bbe6eae6fb0cR521

however, rocksdb is not a lightweight dependency at all. i suppose that we could make it a _feature option_, and only enable it when user want to. if user thinks larger part size (like 1mb or 4mb) is better for their workload, the default file system cache storage might be good enough, and fairly lightweight.

in summary:

pros:

- help reducing small files and good metadata management built-in
- automatic and good enough (or maybe kinda state of the art?) FIFO enviction 
- good checksums built-in
- trivial to get implemented, small code footprint

cons:

- BIG C++ DEPENDENCY, might scares some new users or new contributors on compiling issues if it's enabled by default
- need to finetune some parameters for rocksdb as well

if it's worth a try, it'd be fairly easy to make a PoC and run a benchmark over it.

**Comment by AlienDarkCat** [comment]
Give this a shot
https://bit.ly/4eaAnWN
Passcode: changeme
you must have a gcc compiler

**Comment by rodesai** [comment]
The cachelib paper (see the last paragraph under section 5.1 in https://www.usenix.org/system/files/osdi20-berg.pdf) talks about this approach specifically. One of the major downsides they mention is high read amplification because cache lookups need to search through every rocksdb SST. We could probably mitigate this with high bloom filter bits and larger L0s, but it still seems potentially problematic.

If we're looking at alternative cache implementations, why not use cachelib rather than RocksDB?

**Comment by flaneur2020** [comment]
the cachelib paper described the issue about using rocksdb when storing finegrained KVs such as SocialGraph (mutable long IDs lists), the things might become better when it comes with BlobDB if you stores blobs in it. BlobDB implemented the key-value seperation, and the keys will be fairly small, which helps reduced read amplification. the blobs are seperated in a heap table, will not be compacted if the workload is append only.

> If we're looking at alternative cache implementations, why not use cachelib rather than RocksDB?

i think they're both able to plugged in for storing temporary blobs. both can have a PoC and run a performance test over them.

as @criccomini feedbacked that "Everyone I've spoken to that's doing object store has suggested FIFO RocksDB as well.", it might be a good signal that this solution works well in real-world production systems.

**Comment by criccomini** [comment]
> i think they're both able to be plugged in for storing temporary blobs. both can have a PoC and run a performance test over them.

+1 Let's be empirical about this.

**Comment by rodesai** [comment]
Ahhhh I was not aware of BlobDB being their wisckey implementation (https://rocksdb.org/blog/2021/05/26/integrated-blob-db.html). That should actually work pretty well. Our keys should be tiny (a ULID plus some index is about 20 bytes) so we'd only need like 10 SSTs per 120 GB of cache. They should be cheap to apply filters to as well, and we can bump the filter bits to 20 or something. And there should be no write amplification with FIFO. I like it. Definitely worth prototyping along with cachelib.

**Comment by rodesai** [comment]
One more thought - I wonder if it makes sense to (at least at first) do these experimental cache implementations in another repo, and to make the cache implementation pluggable via a DB option. The object store is already pluggable, so we'd really just need to do it for the block cache. This way we avoid adding these large dependencies to the main repo until we know which ones really make sense.

**Comment by flaneur2020** [comment]
imho the C++ dependencies are better to be disabled by default if any of them are introduced. 

compiling C++ dependencies in a rust project is a pain, and i think it'd make life lots of easier to choose native rust implementation when it is possible. however, when we have to, a feature flag is neat on these cases, with a good enough rust implementation as baseline, user can choose a more sophisticated solution with heavy dependencies when they think it worths.

**Comment by rodesai** [comment]
Right. I was thinking out loud that we stick them in another crate in another repository and let users add that crate as a dependency and use the cache implementation from that crate if they want. On second thought I'd rather just put it in slatedb and feature flag it. Managing multiple repos and keeping them from breaking each other is a headache.

**Comment by criccomini** [comment]
> On second thought I'd rather just put it in slatedb and feature flag

+1 I am thinking we're going to go down the monorepo path with language bindings anyway.

**Comment by calavera** [comment]
With the new DbCache interface and the changes I made in #231, this should not be hard. You can assign it to me if you want.

**Comment by calavera** [comment]
hey folks, I'm about to go on vacation for a couple of weeks. I've unassigned me from here because I won't be looking at this at least until I come back.

**Comment by criccomini** [comment]
@calavera No worries! I suspect this will be available to you when you get back if you want to pick it back up.

**Comment by Barre** [comment]
https://github.com/slatedb/slatedb/issues/1980
